core/modules/ccontrol/UserQueryDropTable.cc

```diff
@@ -0,0 +1,187 @@
+#include "qmeta/Exceptions.h"
+#include "qmeta/QMeta.h"
+#include "sql/SqlConnection.h"
+#include "sql/SqlErrorObject.h"
+#include "util/IterableFormatter.h"
+
+namespace {
+
+LOG_LOGGER getLogger() {
+    static LOG_LOGGER logger = LOG_GET("lsst.qserv.ccontrol.UserQueryDropTable");
+    return logger;
+}
+
+}
+
+namespace lsst {
+namespace qserv {
+namespace ccontrol {
+
+// Constructor
+UserQueryDropTable::UserQueryDropTable(std::shared_ptr<css::CssAccess> const& css,
+                                       std::string const& dbName,
+                                       std::string const& tableName,
+                                       sql::SqlConnection* resultDbConn,
+                                       std::string const& resultTable,
+                                       std::shared_ptr<qmeta::QMeta> const& queryMetadata,
+                                       qmeta::CzarId qMetaCzarId)
+    : _css(css), _dbName(dbName), _tableName(tableName),
+      _resultDbConn(resultDbConn), _resultTable(resultTable),
+      _queryMetadata(queryMetadata), _qMetaCzarId(qMetaCzarId),
+      _qState(UNKNOWN), _messageStore(std::make_shared<qdisp::MessageStore>()),
+      _sessionId(0) {
+}
+
+std::string UserQueryDropTable::getError() const {
+    return std::string();
+}
+
+// Attempt to kill in progress.
+void UserQueryDropTable::kill() {
+}
+
+// Submit or execute the query.
+void UserQueryDropTable::submit() {
+    // Just mark this table in CSS with special status, watcher
+    // will take care of the actual delete process
+
+    LOGF(getLogger(), LOG_LVL_INFO, "going to drop table - %s.%s" % _dbName % _tableName);
+
+    // create result table first, exact schema does not matter but mysql
+    // needs at least one column in table DDL
+    LOGF(getLogger(), LOG_LVL_DEBUG, "creating result table: %s" % _resultTable);
+    std::string sql = "CREATE TABLE " + _resultTable + " (CODE INT)";
+    sql::SqlErrorObject sqlErr;
+    if (not _resultDbConn->runQuery(sql, sqlErr)) {
+        // There is no way to return success if we cannot create result table so just stop here
+        std::string message = "Failed to create result table: " + sqlErr.errMsg();
+        _messageStore->addMessage(-1, 1005, message, MessageSeverity::MSG_ERROR);
+        _qState = ERROR;
+        return;
+    }
+
+    // check current table status, if not READY then fail
+    try {
+        auto statusMap = _css->getTableStatus(_dbName);
+        LOGF(getLogger(), LOG_LVL_DEBUG, "all table status: %s" % util::printable(statusMap));
+        if (statusMap.count(_tableName) != 1) {
+            std::string message = "Unknown table " + _dbName + "." + _tableName;
+            _messageStore->addMessage(-1, 1051, message, MessageSeverity::MSG_ERROR);
+            _qState = ERROR;
+            return;
+        }
+        LOGF(getLogger(), LOG_LVL_DEBUG, "table status: %s" % statusMap[_tableName]);
+        if (statusMap[_tableName] != css::KEY_STATUS_READY) {
```

Comment:
First `if` (I guess you mean line 103) is checking that table name is present in CSS metadata, i.e. that table actually exists. This `if` checks its status which is supposed to be "READY", and issues an error if status is anything different. Table may exists but have a status which is not valid for dropping (e.g. "PENDING_DROP").